File: 品味Java的21種設計模式/dp_java/chainofresponsibility/example5/DepManager.java
package cn.javass.dp.chainofresponsibility.example5;
/**
 * 实现部门经理处理聚餐费用申请的对象
 */
public class DepManager extends Handler {
    public Object handleRequest(RequestModel rm) {
        if (FeeRequestModel.FEE_TYPE.equals(rm.getType())) {
            return handleFeeRequest(rm);
        } else {
            return super.handleRequest(rm);
        }
    }
    private Object handleFeeRequest(RequestModel rm) {
        //先把通用的对象造型回来
        FeeRequestModel frm = (FeeRequestModel)rm;
        String str = "";

        //部门经理的权限只能在1000以内
        if (frm.getFee() < 1000) {
            //为了测试，简单点，只同意小李申请的
            if ("小李".equals(frm.getUser())) {
                str = "部门经理同意" + frm.getUser() + "聚餐费用" + frm.getFee() + "元的请求";
            } else {
                //其他人一律不同意
                str = "部门经理不同意" + frm.getUser() + "聚餐费用" + frm.getFee() + "元的请求";
            }

            return str;
        } else {
            //超过1000，继续传递给级别更高的人处理
            if (this.successor != null) {
                return this.successor.handleRequest(rm);
            }
        }

        return str;
    }
}

File: 品味Java的21種設計模式/dp_cpp/chainofresponsibility/example4/DepManager.cpp
#include "DepManager.h"

namespace cn {
namespace javass {
namespace dp {
namespace chainofresponsibility {
namespace example4 {

std::string DepManager::handleFeeRequest(std::string user, double fee) {
    std::string str = "";

    //部门经理的权限只能在1000以内
    if (fee < 1000) {
        //为了测试，简单点，只同意小李申请的
        if ((new std::string("小李"))->equals(user)) {
            str = "部门经理同意" + user + "聚餐费用" + fee + "元的请求";
        } else {
            //其他人一律不同意
            str = "部门经理不同意" + user + "聚餐费用" + fee + "元的请求";
        }

        return str;
    } else {
        //超过1000，继续传递给级别更高的人处理
        if (this->successor != 0) {
            return this->successor->handleFeeRequest(user, fee);
        }
    }

    return str;
}

bool DepManager::handlePreFeeRequest(std::string user, double requestNum) {
    //部门经理的权限比较小，只能在20000以内
    if (requestNum < 20000) {
        //工作需要嘛，统统同意
        puts("部门经理同意" + user + "预支差旅费用" + requestNum + "元的请求");
        return true;
    } else {
        //超过20000，继续传递给级别更高的人处理
        if (this->successor != 0) {
            return this->successor->handlePreFeeRequest(user, requestNum);
        }
    }

    return false;
}
}
}
}
}
}
